File: shared/src/models/device.rs
// Device-related models

use serde::{Deserialize, Serialize};
use uuid::Uuid;
use chrono::{DateTime, Utc};

/// Represents a paired device
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Device {
    pub id: String,
    pub name: String,
    pub device_type: DeviceType,
    pub platform: String,
    pub os_version: String,
    pub paired_at: DateTime<Utc>,
    pub last_seen: DateTime<Utc>,
    pub is_trusted: bool,
    pub public_key: String,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum DeviceType {
    Android,
    Ios,
    Desktop,
}

/// Device capabilities
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DeviceCapabilities {
    pub supports_gps: bool,
    pub supports_heading: bool,
    pub supports_motion: bool,
    pub supports_notifications: bool,
    pub supports_clipboard: bool,
    pub supports_file_transfer: bool,
    pub max_simultaneous_streams: u32,
    pub supported_codecs: Vec<String>,
}

File: companion/lib/models/device.dart
/// Represents this client (or a remembered client) as the server sees it.
/// Used by the pairing flow and stored locally once paired.
library;

class Device {
  final String id;
  final String name;
  final String deviceType; // 'android' | 'ios' | 'desktop'
  final String? platform;
  final String? osVersion;
  final DateTime? pairedAt;
  final DateTime? lastSeen;
  final bool isTrusted;
  final String? publicKey;

  const Device({
    required this.id,
    required this.name,
    required this.deviceType,
    this.platform,
    this.osVersion,
    this.pairedAt,
    this.lastSeen,
    this.isTrusted = false,
    this.publicKey,
  });

  factory Device.fromJson(Map<String, dynamic> json) {
    return Device(
      id: json['id'] as String,
      name: json['name'] as String,
      deviceType: json['device_type'] as String? ?? 'android',
      platform: json['platform'] as String?,
      osVersion: json['os_version'] as String?,
      pairedAt: json['paired_at'] != null ? DateTime.parse(json['paired_at'] as String) : null,
      lastSeen: json['last_seen'] != null ? DateTime.parse(json['last_seen'] as String) : null,
      isTrusted: json['is_trusted'] as bool? ?? false,
      publicKey: json['public_key'] as String?,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'name': name,
        'device_type': deviceType,
        if (platform != null) 'platform': platform,
        if (osVersion != null) 'os_version': osVersion,
        if (pairedAt != null) 'paired_at': pairedAt!.toIso8601String(),
        if (lastSeen != null) 'last_seen': lastSeen!.toIso8601String(),
        'is_trusted': isTrusted,
        if (publicKey != null) 'public_key': publicKey,
      };
}
